finality-aleph/src/data_io/data_provider.rs

```diff
@@ -280,25 +280,23 @@ where
         }
     }
 
-    pub async fn run(mut self, mut exit: oneshot::Receiver<()>) {
+    async fn main_loop(&mut self) -> Bottom {
         let mut best_block_in_session: Option<BlockId> = None;
         loop {
-            let delay = futures_timer::Delay::new(self.config.refresh_interval);
-            tokio::select! {
-                _ = delay => {
-                    best_block_in_session = self.get_best_block_in_session(best_block_in_session).await;
-                    if let Some(best_block) = &best_block_in_session {
-                        self.update_data(best_block);
-                    }
-
-                }
-                _ = &mut exit => {
-                    debug!(target: "aleph-data-store", "Task for refreshing best chain received exit signal. Terminating.");
-                    return;
-                }
+            futures_timer::Delay::new(self.config.refresh_interval).await;
+            best_block_in_session = self.get_best_block_in_session(best_block_in_session).await;
+            if let Some(best_block) = &best_block_in_session {
+                self.update_data(best_block);
             }
         }
     }
+
+    pub async fn run(mut self, exit: oneshot::Receiver<()>) {
+        tokio::select! {
+            _ = self.main_loop() => error!(target: "aleph-data-store", "Task for refreshing best chain finished."),
```

Comment:
Compiler is not able to deduce that this is unreachable? If it isn't, I think `error` is indeed the best option out of `panic!` and `{}`.

Reply:
Hm, I don't want to keep `Bottom` as a type in general (see another comment in this review), but wouldn't an empty `match` be technically exhaustive, and thus solve this issue? That is:
```
impossible_result = self.main_loop() => match impossible_result {},
```

Reply:
What about using `futures::future::select` ?

Reply:
Looks good